auth/opa/proxiedidentity/proxiedidentity.go

```diff
@@ -0,0 +1,137 @@
+/* Copyright (c) 2023 Snowflake Inc. All rights reserved.
+
+   Licensed under the Apache License, Version 2.0 (the
+   "License"); you may not use this file except in compliance
+   with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing,
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+*/
+
+// Package proxiedidentity provides a way to pass the identity of an end user
+// through the SansShell proxy
+package proxiedidentity
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/Snowflake-Labs/sansshell/auth/opa/rpcauth"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// We don't expose the key so that people aren't tempted to use it directly.
+// We intentionally prefix this with something other than sansshell- so that
+// https://github.com/Snowflake-Labs/sansshell/blob/main/telemetry/telemetry.go
+// doesn't accidentally send untrusted data in passAlongMetadata().
+const reqProxiedIdentityKey = "proxied-sansshell-identity"
+
+// ServerProxiedIdentityUnaryInterceptor adds information about a proxied caller to the RPC context
+// if the provided function returns true. Allow functions will typically pull out information on the
```

Comment:
I'm having trouble here understanding how the 'allow' will work in practice. I don't see a call site other than in tests, where the allow function is trivial.

It feels kinda like an authz check outside the OPA policy, but maybe I'm thinking about it wrong.

Reply:
There's a more concrete example in sansshell-server main in https://github.com/Snowflake-Labs/sansshell/pull/374

```
allowProxyToImpersonate := func(ctx context.Context) bool {
	peer := rpcauth.PeerInputFromContext(ctx)
	if peer == nil {
		return false
	}
	return peer.Cert.Subject.CommonName == *proxyIdentity
}
```

I was considering making it part of the authz policy. It'd look something like

```
package sansshell.authz
default authz = false
authz {
    allow
    not deny
}
deny {
  input.metadata["proxied-sansshell-identity"]
  not input.peer.principal.id = "proxy"
}
```

My worry was that it might be too easy to add the interceptors without updating the rego policy. Now that I think about it more, the benefit of consolidating authz might outweigh that worry.

Reply:
Changed to recommend using the OPA policy, ptal